src/django_components/component.py

```diff
@@ -1022,101 +1085,173 @@ def _render_impl(
         cache_component_css(self.__class__)
         css_input_hash = cache_component_css_vars(self.__class__, css_data) if css_data else None
 
-        with _prepare_template(self, context, context_data) as template:
+        with _prepare_template(self, context, context_data, metadata) as template:
+            component_ctx.template_name = template.name
+
             # For users, we expose boolean variables that they may check
             # to see if given slot was filled, e.g.:
             # `{% if variable > 8 and component_vars.is_filled.header %}`
             is_filled = SlotIsFilled(slots_untyped)
-            self._render_stack[-1].is_filled = is_filled
-
-            # If any slot fills were defined within the template, we want to scope them
-            # to the CSS of the parent component. Thus we keep track of the parent component.
-            if context.get(_COMPONENT_SLOT_CTX_CONTEXT_KEY, None):
-                parent_comp_ctx: ComponentSlotContext = context[_COMPONENT_SLOT_CTX_CONTEXT_KEY]
-                parent_id = parent_comp_ctx.component_id
-            else:
-                parent_id = None
-
-            component_slot_ctx = ComponentSlotContext(
-                component_name=self.name,
-                component_id=render_id,
-                template_name=template.name,
-                fills=slots_untyped,
-                is_dynamic_component=getattr(self, "_is_dynamic_component", False),
-                # This field will be modified from within `SlotNodes.render()`:
-                # - The `default_slot` will be set to the first slot that has the `default` attribute set.
-                #   If multiple slots have the `default` attribute set, yet have different name, then
-                #   we will raise an error.
-                default_slot=None,
-            )
+            metadata.is_filled = is_filled
 
             with context.update(
                 {
                     # Private context fields
-                    _ROOT_CTX_CONTEXT_KEY: self.outer_context,
```

Comment:
I also reduce the number of keys we need to pass down the Context object. Now it's just 2:
- internal `_COMPONENT_CONTEXT_KEY` (aka `"_DJC_COMPONENT_CTX"`)
- public `"component_vars"`